src/main/java/org/victorrobotics/frc/dtlib/drivetrain/swerve/trajectory/DTSwerveTrajectoryGenerator.java

```diff
@@ -0,0 +1,280 @@
+
+    private DTAccelerationLimit accelerationLimit;
+    private DTVelocityLimit     velocityLimit;
+
+    private double distanceBetweenPoints;
+    private double smoothWeight;
+    private double velocityPower;
+    private double velocityCoefficient;
+
+    public DTSwerveTrajectoryGenerator() {
+        distanceBetweenPoints = DEFAULT_DISTANCE_BETWEEN_POINTS;
+        smoothWeight = DEFAULT_SMOOTH_WEIGHT;
+        velocityPower = DEFAULT_VELOCITY_POWER;
+        velocityCoefficient = DEFAULT_VELOCITY_COEFFICIENT;
+
+        accelerationLimit = DEFAULT_ACCEL_LIMIT;
+        velocityLimit = DEFAULT_VELO_LIMIT;
+    }
+
+    public void setAccelerationLimit(DTAccelerationLimit limit) {
+        accelerationLimit = limit;
+    }
+
+    public void setVelocityLimit(DTVelocityLimit limit) {
+        velocityLimit = limit;
+    }
+
+    public void setDistanceBetweenPoints(double distanceMeters) {
+        distanceBetweenPoints = distanceMeters;
+    }
+
+    public void setSmoothWeight(double weight) {
+        smoothWeight = weight;
+    }
+
+    public void setVelocityPower(double power) {
+        velocityPower = power;
+    }
+
+    public void setVelocityCoefficient(double coefficient) {
+        velocityCoefficient = coefficient;
+    }
+
+    public DTSwerveTrajectory generate(Pose2d... waypoints) {
+        double[][] interpolatedPoints = interpolate(waypoints);
+        double[][] smoothPoints = smooth(interpolatedPoints);
+
+        Pose2d[] path = new Pose2d[smoothPoints.length];
+        for (int i = 0; i < path.length; i++) {
+            path[i] = new Pose2d(smoothPoints[i][0], smoothPoints[i][1],
+                    Rotation2d.fromDegrees(smoothPoints[i][2]));
+        }
+
+        // cumulative time, cumulative distance, commanded velocity, predicted
+        // velocity, linear acceleration, curvature
+        double[][] dataValues = new double[6][path.length];
+        computeVelocity(path, dataValues[1], dataValues[2], dataValues[3], dataValues[5]);
+        computeAcceleration(dataValues[1], dataValues[3], dataValues[0], dataValues[4]);
+
+        Point[] points = constructPoints(path, dataValues[0], dataValues[1], dataValues[2],
+                dataValues[3], dataValues[4], dataValues[5]);
+        return new DTSwerveTrajectory(points);
+    }
+
+    private double[][] interpolate(Pose2d[] waypoints) {
+        // Calculate how many points to insert into each segment and their
+        // initial transforms
+        int[] intermediatePointCounts = new int[waypoints.length - 1];
+        double[][] deltas = new double[waypoints.length - 1][3];
+        int totalPointCount = 1;
+        for (int i = 0; i < deltas.length; i++) {
+            Pose2d currentPoint = waypoints[i];
+            Pose2d nextPoint = waypoints[i + 1];
+            deltas[i][0] = nextPoint.getX() - currentPoint.getX();
+            deltas[i][1] = nextPoint.getY() - currentPoint.getY();
+            deltas[i][2] = nextPoint.getRotation()
```

Comment:
You could do the same here. Alternatively, make local variables `dx, dy, dr` to hold what is currently `deltas[i][...]` until you've finished calculating them, then assign to `deltas[i][...]` at the end.